soufi/finders/gem.py

```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2021 Cisco Systems, Inc. and its affiliates
+# All rights reserved.
+
+import requests
+
+from soufi import exceptions, finder
+
+GEM_DOWNLOADS = 'https://rubygems.org/downloads/'
+API_TIMEOUT = 30  # seconds
+
+
+class GemFinder(finder.SourceFinder):
+    """Find Gem files.
+
+    Retrieves files from https://rubygems.org/downloads
+    """
+
+    distro = finder.SourceType.gem.value
+
+    def _find(self):
+        source_url = self.get_source_url()
+        return GemDiscoveredSource([source_url])
+
+    def get_source_url(self):
+        url = f"{GEM_DOWNLOADS}{self.name}-{self.version}.gem"
+        response = requests.head(url, timeout=API_TIMEOUT)
+        if response.status_code != requests.codes.ok:
+            raise exceptions.SourceNotFound
+        return url
+
+
+class GemDiscoveredSource(finder.DiscoveredSource):
+    """A discovered Gem package."""
+
+    # It's already an archive, and it's already properly named
+    make_archive = finder.DiscoveredSource.remote_url_is_archive
+    archive_extension = ''
```

Comment:
We need to set this because it's part of the external API contract. I realise that means you will end up with `foo.gem.gem` on the auto-files but you can override that if you re-implement the archive making part if you want. Otherwise .gem.gem is not the end of the world.